common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +173,445 @@
+                        instance = holder.get();
+                        if (instance == null) {
+                            instance = createNewExtension(definition, loader, argTypes, args);
+                            holder.set(instance);
+                        }
+                    }
+                }
+                return (S)instance;
             } else {
-                throw new EnhancedServiceNotFoundException(
-                    "not found service provider for : " + service.getName() + " caused by " + ExceptionUtils
-                        .getFullStackTrace(e));
+                return createNewExtension(definition, loader, argTypes, args);
             }
         }
-    }
 
-    @SuppressWarnings("rawtypes")
-    private static <S> List<Class> findAllExtensionClass(Class<S> service, String activateName, ClassLoader loader) {
-        List<Class> extensions = new ArrayList<>();
-        try {
-            loadFile(service, SERVICES_DIRECTORY, loader, extensions);
-            loadFile(service, SEATA_DIRECTORY, loader, extensions);
-        } catch (IOException e) {
-            throw new EnhancedServiceNotFoundException(e);
+        private S createNewExtension(ExtensionDefinition definition, ClassLoader loader, Class[] argTypes, Object[] args) {
+            loadAllExtensionClass(loader);
+            Class<?> clazz = definition.getServiceClass();
+            try {
+                S newInstance = initInstance(clazz, argTypes, args);
+                return newInstance;
+            } catch (Throwable t) {
+                throw new IllegalStateException("Extension instance(definition: " + definition + ", class: " +
+                        type + ")  could not be instantiated: " + t.getMessage(), t);
+            }
         }
 
-        if (extensions.isEmpty()) {
-            return extensions;
-        }
-        extensions.sort((c1, c2) -> {
-            int o1 = 0;
-            int o2 = 0;
-            @SuppressWarnings("unchecked")
-            LoadLevel a1 = (LoadLevel) c1.getAnnotation(LoadLevel.class);
-            @SuppressWarnings("unchecked")
-            LoadLevel a2 = (LoadLevel) c2.getAnnotation(LoadLevel.class);
-
-            if (a1 != null) {
-                o1 = a1.order();
+        private List<Class> loadAllExtensionClass(ClassLoader loader) {
+            List<Class> result;
+            List<ExtensionDefinition> definitions = definitionsHolder.get();
+            if (definitions == null) {
+                synchronized (definitionsHolder) {
+                    definitions = definitionsHolder.get();
+                    if (definitions == null) {
+                        definitions = findAllExtensionDefinition(loader);
+                        definitionsHolder.set(definitions);
+                    }
+                }
             }
+            result = definitions.stream().map(def -> def.getServiceClass()).collect(Collectors.toList());
+            return result;
```

Comment:
Remove var `result` and `return definitions.stream().map(ExtensionDefinition::getServiceClass).collect(Collectors.toList())`?

Reply:
OK, done